schema/schema_registry.cc

```diff
@@ -54,6 +54,19 @@ void schema_registry::init(const db::schema_ctxt& ctxt) {
     _ctxt = std::make_unique<db::schema_ctxt>(ctxt);
 }
 
+void schema_registry::attach_table(schema_registry_entry& e) noexcept {
+    replica::database* db = _ctxt->get_db();
+    if (!db) {
+        return;
+    }
+    try {
+        auto& table = db->find_column_family(e.get_schema()->id());
+        e.set_table(table.weak_from_this());
+    } catch (const replica::no_such_column_family&) {
+        // ignore
+    }
+}
+
```

Comment:
I'm fairly negative on this. The schema should be just a schema. What are those contexts that only have a schema and don't have the database?

I'll try to suggest alternatives as I review.

Reply:
It will also be useful to get rid of the "static props" thing added in this PR: https://github.com/scylladb/scylladb/pull/13170#issuecomment-1469848086

Reply:
It's truly necessary only to add a check to schema::get_sharder() which catches sites which use it with tablet-based tables.

Other uses could either lookup in the database or propagate erm across layer.

There is some advantage in that we avoid table lookups, which would have to happen multiple times at different layers.

Reply:
I believe most call sites can be adjusted to use an already-looked-up table. Or outright eliminated, like multishard_mutation_query. However I don't propose doing that as a prerequisite, we won't make progress.

> cdc/metadata.cc:    auto shard_id = dht::shard_of(shard_cnt, entry.sharding_ignore_msb, tok);


cdc sharding will be reworked for (and greatly benfit from) tablets

> compaction/compaction.cc:        auto shard = dht::shard_of(*_schema, dk.token());

This is for resharding compactions, which are unnecessary with compaction groups.

> cql3/statements/select_statement.cc:        unsigned shard = dht::shard_of(*_schema, key_ranges[0].start()->value().as_decorated_key().token());

has access to global context.

> db/commitlog/commitlog_replayer.cc:        auto shard = fm.shard_of(schema);

Can be delegated access to global context.


> db/size_estimates_virtual_reader.cc:            return shard_of(s, dht::get_token(s,

Is in table context

> db/virtual_table.cc:    return dht::shard_of(*_s, dk.token()) == this_shard_id();

Is in table context


> dht/i_partitioner.cc:sharder::shard_of(const token& t) const {
> dht/i_partitioner.cc:    return dht::shard_of(_shard_count, _sharding_ignore_msb_bits, t);
> dht/i_partitioner.cc:unsigned shard_of(const schema& s, const token& t) {
> dht/i_partitioner.cc:    return s.get_sharder().shard_of(t);
> dht/i_partitioner.cc:    auto shard = _range.start() ? _sharder.shard_of(_range.start()->value().token()) : token::shard_of_minimum_token();
> dht/i_partitioner.cc:    auto start_shard = sharder.shard_of(start_token);
> dht/i_partitioner.hh:unsigned shard_of(const schema&, const token&);
> dht/sharder.hh:            , _start_boundary(_sharder.shard_of(_start_token) == shard ?
> dht/token-sharding.hh:unsigned shard_of(unsigned shard_count, unsigned sharding_ignore_msb_bits, const token& t);
> dht/token-sharding.hh:    virtual unsigned shard_of(const token& t) const;
> dht/token.cc:shard_of(unsigned shard_count, unsigned sharding_ignore_msb_bits, const token& t) {
> dht/token.cc:    if (sharder.shard_of(t) != shard_idx) {


This is the infrastructure we're eliminating

> docs/dev/cdc.md:_entries[i][p.shard_of(tok, _entries[i].streams.size(), _entries[i].sharding_ignore_msb)]
> Binary file docs/kb/flamegraph.svg matches

Ignorable

> multishard_mutation_query.cc:            const auto shard = sharder.shard_of(rit->as_partition_start().key().token());
> multishard_mutation_query.cc:    const auto shard = sharder.shard_of(pkey.token());
> multishard_mutation_query.cc:    const auto shard = 
sharder.shard_of(compaction_state.partition_start.key().token());

Irrelevant for tablets

> mutation/frozen_mutation.hh:    unsigned shard_of(const schema& s) const {
> mutation/frozen_mutation.hh:        return dht::shard_of(s, dht::get_token(s, key()));
> mutation/mutation.hh:    unsigned shard_of() const {
> mutation/mutation.hh:        return dht::shard_of(*schema(), token());

To be eliminated

> mutation_writer/multishard_writer.cc:        return _s->get_sharder().shard_of(mf.as_partition_start().key().token());
> mutation_writer/shard_based_splitting_writer.cc:        _current_shard = dht::shard_of(*_schema, ps.key().token());
> readers/multishard.cc:    _current_shard = _sharder.shard_of(token);


Irrelevant for tablets


> replica/database.cc:        auto m_shard = dht::shard_of(*s, dht::get_token(*s, muts[i].key()));
> replica/database.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> replica/database.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> replica/database.hh://   use shard_of() for data
> replica/database.hh:    // All mutations must be owned by the current shard (in terms of dht::shard_of).
> replica/table.cc:        if (dht::shard_of(*s, pos.token()) != this_shard_id()) {

Have access to global context 

> scylla-gdb.py:def shard_of(ptr):
> scylla-gdb.py:        shard = shard_of(cache_region.impl())
> scylla-gdb.py:                if shard_of(int(desc.address)) != shard:

Irrelevant

> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:    return dht::shard_of(s, token);
> service/storage_proxy.cc:    if (coordinator_in_replica_set && dht::shard_of(*s, token) == this_shard_id()) {
> service/storage_proxy.cc:    auto shard = m.shard_of();
> service/storage_proxy.cc:    auto shard = m.shard_of(*s);
> service/storage_proxy.cc:    auto shard = m.shard_of(*s);
> service/storage_proxy.cc:    auto shard = fm.shard_of(*s);
> service/storage_proxy.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> service/storage_proxy.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());

Have access to global context

> test/boost/commitlog_test.cc:            return m.shard_of();
> test/boost/database_test.cc:    auto shard = m.shard_of();
> test/boost/database_test.cc:            auto shard = m.shard_of();
> test/boost/database_test.cc:                auto shard = m.shard_of();
> test/boost/database_test.cc:                auto shard = m.shard_of();
> test/boost/multishard_mutation_query_test.cc:                return smp::submit_to(dht::shard_of(*schema, mut.decorated_key().token()), [&env, gs = global_schema_ptr(schema), mut = freeze(mut)] () mutable {
> test/boost/multishard_mutation_query_test.cc:        const auto res = db.invoke_on(sharder.shard_of(pkey.token()), [gs = global_schema_ptr(s), &pkey, &slice] (replica::database& db) {
> test/boost/mutation_reader_test.cc:        auto shard = sharder.shard_of(start_token);
> test/boost/mutation_reader_test.cc:                    return remote_partitioner.shard_of(pkey.token()) == 0;
> test/boost/mutation_writer_test.cc:                    auto shard = s->get_sharder().shard_of(m.token());
> test/boost/mutation_writer_test.cc:                                        auto shard = sharder.shard_of(mf_opt->as_partition_start().key().token());
> test/boost/mutation_writer_test.cc:                                        auto shard = sharder.shard_of(mf_opt->as_partition_start().key().token());
> test/boost/partitioner_test.cc:        BOOST_REQUIRE_EQUAL(sharder.shard_of(lim), i % shards);
> test/boost/partitioner_test.cc:            BOOST_REQUIRE_EQUAL(sharder.shard_of(prev_token(lim)), (i - 1) % shards);
> test/boost/partitioner_test.cc:            BOOST_REQUIRE_EQUAL(sharder.shard_of(sharder.token_for_next_shard(lim, next_shard)), next_shard);
> test/boost/partitioner_test.cc:                auto start_shard = input_sharder.shard_of(range_shard->start()->value());
> test/boost/partitioner_test.cc:                auto end_shard = input_sharder.shard_of(range_shard->end()->value());
> test/boost/partitioner_test.cc:            auto mid_shard = input_sharder.shard_of(midpoint);
> test/boost/partitioner_test.cc:    auto last_shard = sharder.shard_of(last_token);
> test/boost/per_partition_rate_limit_test.cc:        unsigned local_shard = dht::shard_of(*sptr, dht::get_token(*sptr, pk.view()));
> test/boost/secondary_index_test.cc:                unsigned shard = m.shard_of();
> test/lib/cql_test_env.cc:        auto shard = dht::shard_of(*schema, dk._token);
> test/lib/dummy_sharder.cc:unsigned dummy_sharder::shard_of(const dht::token& t) const {
> test/lib/dummy_sharder.cc:        return shard_token > t && shard_of(shard_token) == shard;
> test/lib/dummy_sharder.hh:    virtual unsigned shard_of(const dht::token& t) const override;
> test/lib/key_utils.cc:                if (shard && *shard != dht::shard_of(*s, dkey.token())) {


Fixing up the tests will be painful

> tools/scylla-types.cc:            auto shard = dht::shard_of(vm["shards"].as<unsigned>(), vm["ignore-msb-bits"].as<unsigned>(), dk.token());

Will be broken by tablets (but replaces with a query into the tablets table).